Lib/xml/etree/ElementTree.py

```diff
@@ -836,63 +836,81 @@ def _get_writer(file_or_filename, encoding):
                     if prefix != "xml":
                         namespaces[uri] = prefix
-                if prefix:
-                    qnames[qname] = "%s:%s" % (prefix, tag)
-                else:
-                    qnames[qname] = tag # default element
+            if prefix:
+                return "%s:%s" % (prefix, local)
             else:
-                if default_namespace:
-                    # FIXME: can this be handled in XML 1.0?
-                    raise ValueError(
-                        "cannot use non-qualified names with "
-                        "default_namespace option"
-                        )
-                qnames[qname] = qname
-        except TypeError:
-            _raise_serialization_error(qname)
+                return local # default element
+        else:
+            if not is_attr and default_namespace:
+                # FIXME: can this be handled in XML 1.0?
+                raise ValueError(
+                    "cannot use non-qualified name (<%s>) with "
+                    "default_namespace option" % qname
+                    )
+            return qname
+
+    def add_qname(qname, is_attr=False):
+        ser_tag, ser_attr = qnames.get(qname, (None, None))
+        if is_attr:
+            if ser_attr is None:
+                ser_attr = serialize_qname(qname, True)
+                if not default_namespace:
+                    ser_tag = ser_attr
+                qnames[qname] = (ser_tag, ser_attr)
+        else:
+            if ser_tag is None:
+                ser_tag = serialize_qname(qname, False)
+                if not default_namespace:
+                    ser_attr = ser_tag
+                qnames[qname] = (ser_tag, ser_attr)
 
     # populate qname and namespaces table
     for elem in elem.iter():
         tag = elem.tag
         if isinstance(tag, QName):
-            if tag.text not in qnames:
-                add_qname(tag.text)
+            add_qname(tag.text)
         elif isinstance(tag, str):
-            if tag not in qnames:
-                add_qname(tag)
+            add_qname(tag)
         elif tag is not None and tag is not Comment and tag is not PI:
             _raise_serialization_error(tag)
         for key, value in elem.items():
             if isinstance(key, QName):
                 key = key.text
-            if key not in qnames:
-                add_qname(key)
-            if isinstance(value, QName) and value.text not in qnames:
+            elif not isinstance(key, str):
+                _raise_serialization_error(key)
+            add_qname(key, is_attr=True)
+            if isinstance(value, QName):
                 add_qname(value.text)
         text = elem.text
-        if isinstance(text, QName) and text.text not in qnames:
+        if isinstance(text, QName):
             add_qname(text.text)
-    return qnames, namespaces
+
+    prefix_map = {prefix: ns for ns, prefix in namespaces.items()}
```

Comment:
Prefixes don't have to be globally unique, so this might introduce bugs. (Suggests to me that there might be missing tests.)